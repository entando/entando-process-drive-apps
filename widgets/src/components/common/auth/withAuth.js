import React, { useState, useEffect } from 'react';
import { IS_MOCKED_API, LOCAL } from 'api/constants';
import utils from 'utils';
import Loader from 'components/common/auth/Loader';
import MissingPermissions from 'components/common/auth/MissingPermissions';

const detectKeycloak = () => window && window.entando && window.entando.keycloak;

const withAuth = (WrappedComponent, permissions = []) => props => {
  const [authLoaded, setAuthLoaded] = useState(false);
  const [missingPermissions, setMissingPermissions] = useState([]);
  const [userRoles, setUserRoles] = useState([]);
  useEffect(() => {
    const beginCheck = async () => {
      await utils.timeout(100);
      const keycloak = detectKeycloak();
      if (LOCAL || IS_MOCKED_API || (!keycloak && localStorage.getItem('token'))) {
        loadPermissions();
      } else {
        if (keycloak.authenticated) {
          loadPermissions(keycloak);
        } else {
          window.addEventListener('keycloak', ({ detail: { eventType } }) => {
            if (eventType === 'onInit' && keycloak.authenticated) {
              loadPermissions(keycloak);
            }
          });
        }
      }
    };
    beginCheck();
  }, []);

  const loadPermissions = keycloak => {
    if (keycloak) {
      const roles =
        (keycloak.clientId &&
          keycloak.resourceAccess &&
          keycloak.resourceAccess[keycloak.clientId] &&
          keycloak.resourceAccess[keycloak.clientId].roles) ||
        [];
      setMissingPermissions(
        permissions.filter(neededPermission => !roles.includes(neededPermission))
      );
      setUserRoles(roles);
    }
    setAuthLoaded(true);
  };

  if (!authLoaded) {
    return <Loader />;
  }
  if (missingPermissions.length > 0) {
    return <MissingPermissions missingPermissions={missingPermissions} />;
  }
  return <WrappedComponent {...props} userRoles={userRoles} />;
};

export default withAuth;
