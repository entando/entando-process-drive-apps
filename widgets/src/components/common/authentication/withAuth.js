import React from 'react';

import Loader from 'components/common/authentication/Loader';
import NotAuthenticated from 'components/common/authentication/NotAuthenticated';
import MissingPermissions from 'components/common/authentication/MissingPermissions';

export default function withAuth(AuthenticatedComponent, permissions = []) {
  return class AuthenticationWrapper extends React.Component {
    static getKeycloak() {
      return (window && window.entando && window.entando.keycloak) || null;
    }

    constructor(props) {
      super(props);

      this.state = {
        initialized: false, // keycloak is ready
        authenticated: false, // user is logged-in
        authorized: false, // user has correct access rights (roles)
        missingPermissions: [],
      };

      this.handleKeycloakEvent = this.handleKeycloakEvent.bind(this);
      this.manageAuthentication = this.manageAuthentication.bind(this);
    }

    componentDidMount() {
      window.addEventListener('keycloak', this.handleKeycloakEvent);

      this.manageAuthentication();
    }

    componentWillUnmount() {
      window.removeEventListener('keycloak', this.handleKeycloakEvent);
    }

    handleKeycloakEvent() {
      this.manageAuthentication();
    }

    manageAuthentication() {
      const keycloak = AuthenticationWrapper.getKeycloak();

      if (keycloak) {
        if (!keycloak.authenticated) {
          this.setState({
            initialized: true,
            authenticated: keycloak.authenticated || false,
          });
        } else {
          const userRoles =
            (keycloak.clientId &&
              keycloak.resourceAccess &&
              keycloak.resourceAccess[keycloak.clientId] &&
              keycloak.resourceAccess[keycloak.clientId].roles) ||
            [];
          const missingPermissions = permissions.filter(
            neededPermission => !userRoles.includes(neededPermission)
          );

          this.setState({
            initialized: true,
            authenticated: keycloak.authenticated || false,
            missingPermissions,
          });
        }
      }
    }

    render() {
      const { authenticated, initialized, missingPermissions } = this.state;

      // if there are no mandatory permissions, component can be displayed as is
      if (permissions.length === 0 || process.env.REACT_APP_DISABLE_KEYCLOAK === 'true') {
        // eslint-disable-next-line react/jsx-props-no-spreading
        return <AuthenticatedComponent {...this.props} />;
      }

      if (permissions.length > 0) {
        // if keycloak is has not been initialized yet
        if (!initialized) {
          return <Loader />;
        }

        if (!authenticated) {
          return <NotAuthenticated />;
        }

        if (missingPermissions.length > 0) {
          return <MissingPermissions missingPermissions={missingPermissions} />;
        }

        // eslint-disable-next-line react/jsx-props-no-spreading
        return <AuthenticatedComponent {...this.props} />;
      }

      return null;
    }
  };
}
