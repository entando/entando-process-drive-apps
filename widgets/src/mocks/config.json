{
  "payload": {
    "code": "pda_task_list",
    "config": {
      "groups": "[{\"label\":\"Administrators\",\"key\":\"Administrators\",\"checked\":true},{\"label\":\"broker\",\"key\":\"broker\",\"checked\":true},{\"label\":\"admin\",\"key\":\"admin\",\"checked\":false},{\"label\":\"rest-all\",\"key\":\"rest-all\",\"checked\":false},{\"label\":\"kiemgmt\",\"key\":\"kiemgmt\",\"checked\":true},{\"label\":\"kie-server\",\"key\":\"kie-server\",\"checked\":true},{\"label\":\"appraiser\",\"key\":\"appraiser\",\"checked\":false},{\"label\":\"manager\",\"key\":\"manager\",\"checked\":true},{\"label\":\"supplier\",\"key\":\"supplier\",\"checked\":true},{\"label\":\"IT\",\"key\":\"IT\",\"checked\":true},{\"label\":\"PM\",\"key\":\"PM\",\"checked\":true}]",
      "options": "[{\"key\":\"newPageOnClick\",\"label\":\"Open new page on table row click\",\"checked\":false},{\"key\":\"showClaim\",\"label\":\"Show Claim Button\",\"checked\":true},{\"key\":\"showComplete\",\"label\":\"Show Complete Button\",\"checked\":true}]",
      "columns": "[{\"name\":\"task-is-skipable\",\"position\":0,\"isVisible\":false},{\"name\":\"task-parent-id\",\"position\":1,\"isVisible\":false},{\"name\":\"task-created-by\",\"position\":2,\"isVisible\":true},{\"name\":\"task-input-data.application.property.saleprice\",\"position\":3,\"isVisible\":true},{\"name\":\"task-input-data.TaskName\",\"position\":4,\"isVisible\":true},{\"name\":\"initiator\",\"position\":5,\"isVisible\":true},{\"name\":\"task-created-on\",\"position\":6,\"isVisible\":true},{\"name\":\"task-priority\",\"position\":7,\"isVisible\":false},{\"name\":\"task-proc-def-id\",\"position\":8,\"isVisible\":true},{\"name\":\"task-form\",\"position\":9,\"isVisible\":true},{\"name\":\"task-skippable\",\"position\":10,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.name\",\"position\":11,\"isVisible\":true},{\"name\":\"task-actual-owner\",\"position\":12,\"isVisible\":true},{\"name\":\"task-activation-time\",\"position\":13,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.address\",\"position\":14,\"isVisible\":true},{\"name\":\"task-input-data.application.errors\",\"position\":15,\"isVisible\":true},{\"name\":\"task-name\",\"position\":16,\"isVisible\":true},{\"name\":\"task-input-data.application.property.address\",\"position\":17,\"isVisible\":true},{\"name\":\"task-subject\",\"position\":18,\"isVisible\":true},{\"name\":\"containerId\",\"position\":19,\"isVisible\":true},{\"name\":\"task-container-id\",\"position\":20,\"isVisible\":true},{\"name\":\"processDefinitionId\",\"position\":21,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.ssn\",\"position\":22,\"isVisible\":true},{\"name\":\"task-description\",\"position\":23,\"isVisible\":true},{\"name\":\"task-input-data.Skippable\",\"position\":24,\"isVisible\":true},{\"name\":\"task-workitem-id\",\"position\":25,\"isVisible\":true},{\"name\":\"task-expiration-time\",\"position\":26,\"isVisible\":true},{\"name\":\"task-input-data.application.property.age\",\"position\":27,\"isVisible\":true},{\"name\":\"task-status\",\"position\":28,\"isVisible\":true},{\"name\":\"task-input-data.application.mortgageamount\",\"position\":29,\"isVisible\":true},{\"name\":\"task-process-instance-id\",\"position\":30,\"isVisible\":true},{\"name\":\"task-input-data.ActorId\",\"position\":31,\"isVisible\":true},{\"name\":\"task-process-id\",\"position\":32,\"isVisible\":true},{\"name\":\"task-input-data.application.property.locale\",\"position\":33,\"isVisible\":true},{\"name\":\"task-input-data.application.amortization\",\"position\":34,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.creditrating\",\"position\":35,\"isVisible\":true},{\"name\":\"application\",\"position\":36,\"isVisible\":true},{\"name\":\"task-input-data.NodeName\",\"position\":37,\"isVisible\":true},{\"name\":\"task-proc-inst-id\",\"position\":38,\"isVisible\":true},{\"name\":\"task-input-data.application.downpayment\",\"position\":39,\"isVisible\":true},{\"name\":\"task-input-data.application.applicant.annualincome\",\"position\":40,\"isVisible\":true},{\"name\":\"task-id\",\"position\":41,\"isVisible\":true}]",
      "knowledgeSource": "kieStaging",
      "process": "itorders.orderhardware@itorders_1.0.0-SNAPSHOT"
    }
  },
  "metaData": {
    "status": "draft"
  },
  "errors": []
}
