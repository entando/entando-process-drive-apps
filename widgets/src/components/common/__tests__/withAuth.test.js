import React from 'react';
import PropTypes from 'prop-types';
import { render, screen } from '@testing-library/react';

import 'mocks/i18nMock';
import withAuth from 'components/common/authentication/withAuth';
import mockKeycloak from 'mocks/auth/keycloak';

const AuthenticatedComponent = ({ value }) => <div>{value}</div>;
AuthenticatedComponent.propTypes = { value: PropTypes.string.isRequired };

// beforeEach(() => {
//   window.entando = {};
// });

describe('withAuth()', () => {
  it('passes props through', async () => {
    const ComponentWithAuth = withAuth(AuthenticatedComponent);

    const { container } = render(<ComponentWithAuth value="Some value" />);

    expect(container).toMatchSnapshot();
  });

  it('shows loading screen when keycloak is not initialized', async () => {
    const ComponentWithAuth = withAuth(AuthenticatedComponent, ['unmet-permission']);

    const { container } = render(<ComponentWithAuth value="Some value" />);

    expect(container).toMatchSnapshot();
  });

  it('shows that permissions are not met', async () => {
    mockKeycloak();
    const ComponentWithAuth = withAuth(AuthenticatedComponent, ['unmet-permission']);

    render(<ComponentWithAuth value="Some value" />);

    expect(
      screen.getByText('authentication.missingPermissions: unmet-permission.')
    ).toBeInTheDocument();
  });

  it('shows widget when keycloak is initialized and permissions are met', async () => {
    mockKeycloak();
    const ComponentWithAuth = withAuth(AuthenticatedComponent, ['task-list']);

    const { container } = render(<ComponentWithAuth value="Some value" />);

    expect(container).toMatchSnapshot();
  });
});
