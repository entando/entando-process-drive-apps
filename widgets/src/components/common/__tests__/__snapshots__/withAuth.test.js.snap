// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withAuth() passes props through 1`] = `
<div>
  <div>
    Some value
  </div>
</div>
`;

exports[`withAuth() shows loading screen when keycloak is not initialized 1`] = `
<div>
  <div
    class="MuiContainer-root Loader-wrapper-1 MuiContainer-disableGutters MuiContainer-maxWidthLg"
  >
    <div
      class="MuiCircularProgress-root WithStyles(ForwardRef(CircularProgress))-root-10 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`withAuth() shows widget when keycloak is initialized and permissions are met 1`] = `
<div>
  <div>
    Some value
  </div>
</div>
`;
